src/common/cockpitjsonprint.c

```diff
@@ -0,0 +1,158 @@
+ */
+
+#define _GNU_SOURCE
+
+#include "cockpitjsonprint.h"
+
+#include <string.h>
+#include <inttypes.h>
+#include <sys/mman.h>
+#include <assert.h>
+#include <fcntl.h>
+
+static bool
+char_needs_json_escape (char c)
+{
+  return c < ' ' || c == '\\' || c == '"';
+}
+
+static bool
+json_escape_char (FILE *stream,
+                  char c)
+{
+  if (c == '\\')
+    return fputs ("\\\\", stream) >= 0;
+  else if (c == '"')
+    return fputs ("\\\"", stream) >= 0;
+  else
+    return fprintf (stream, "\\u%04x", c) == 6;
+}
+
+static bool
+json_escape_string (FILE       *stream,
+                    const char *str,
+                    size_t      maxlen)
+{
+  size_t offset = 0;
+
+  while (offset < maxlen && str[offset])
+    {
+      size_t start = offset;
+
+      while (offset < maxlen && str[offset] && !char_needs_json_escape (str[offset]))
+        offset++;
+
+      /* print the non-escaped prefix, if there is one */
+      if (offset != start)
+        {
+          size_t length = offset - start;
+          if (fwrite (str + start, 1, length, stream) != length)
+            return false;
+        }
+
+      /* print the escaped character, if there is one */
+      if (offset < maxlen && str[offset])
+        {
+          if (!json_escape_char (stream, str[offset]))
+            return false;
+
+          offset++;
+        }
+    }
+
+  return true;
+}
+
+bool
+cockpit_json_print_string_property (FILE       *stream,
+                                    const char *key,
+                                    const char *value,
+                                    ssize_t     maxlen)
+{
+  size_t expected = strlen (key) + 7;
+
+  return fprintf (stream, ", \"%s\": \"", key) == expected &&
+         json_escape_string (stream, value, maxlen) &&
+         fputc ('"', stream) >= 0;
+}
+
+bool
+cockpit_json_print_bool_property (FILE       *stream,
+                                  const char *key,
+                                  bool        value)
+{
+  size_t expected = 6 + strlen (key) + (value ? 4 : 5); /* "true" or "false" */
+
+  return fprintf (stream, ", \"%s\": %s", key, value ? "true" : "false") == expected;
+}
+
+bool
+cockpit_json_print_integer_property (FILE       *stream,
+                                     const char *key,
+                                     uint64_t    value)
+{
+  /* too much effort to figure out the expected length exactly */
+  return fprintf (stream, ", \"%s\": %"PRIu64, key, value) > 6;
+}
+
+FILE *
+cockpit_json_print_open_memfd (const char *name,
+                               int         version)
+{
+  int fd = memfd_create ("cockpit login messages", MFD_ALLOW_SEALING | MFD_CLOEXEC);
+  assert (fd != -1);
+
+  FILE *stream = fdopen (fd, "w");
+
+  if (version >= 0)
+    fprintf (stream, "{\"version\": %d", version);
+  else
+    fputc ('{', stream);
+
+  return stream;
+}
+
+int
+cockpit_json_print_seal_memfd (FILE **stream)
+{
+  int r = fputc ('}', *stream);
+  assert (r == '}');
```

Comment:
I appreciate the simple/robust interface of both closing the object with `}` and sealing the memfd, and closing the file. But "_seal_memfd()" as a function name is then a bit confusing. Maybe `cockpit_json_print_finish_memfd()`? Plus a docstring what it does?

Reply:
To be honest, this function has been inaccurately named since I added the code for reopening the fd readonly.  I'll take your suggested name and add some docs.